biz.aQute.bndlib/src/aQute/bnd/build/WorkspaceRepositoryDynamic.java

```diff
@@ -39,11 +40,16 @@
 	}
 
 	private List<Capability> findProvider(Collection<? extends Resource> resources, Requirement requirement) {
+		List<Capability> l = new ArrayList<>();
 		String namespace = requirement.getNamespace();
-		return resources.stream()
-			.flatMap(resource -> ResourceUtils.capabilityStream(resource, namespace))
-			.filter(ResourceUtils.matcher(requirement))
-			.collect(ResourceUtils.toCapabilities());
+		for (Resource r : resources) {
+			for (Capability c : r.getCapabilities(namespace)) {
+				if (ResourceUtils.matches(requirement, c)) {
```

Comment:
Well you lost an nice performance fix of using `ResourceUtils.matcher(requirement)`. `requirement` never changes in the loop, but `ResourceUtils.matches(requirement, c)` keeps recomputing the filter object every time through the O(n^2) loop. So there was reason for using stream and blindly changing them back can loose things.

Reply:
The code had become completely unreadable. I'll check this optimization. The optimization has got nothing to do with streams.

Reply:
I already have the fix in a commit ready to push.